service/src/test/java/bio/terra/workspace/app/configuration/external/controller/FolderApiControllerTest.java

```diff
@@ -0,0 +1,545 @@
+    // `WorkspaceActivityLogHook` where we extract the user request information and log it to
+    // change agents.
+    when(mockSamService.getUserStatusInfo(any()))
+        .thenReturn(
+            new UserStatusInfo()
+                .userEmail(USER_REQUEST.getEmail())
+                .userSubjectId(USER_REQUEST.getSubjectId()));
+  }
+
+  @Test
+  public void createFolder_parentFolderIdIsNull_topLevelFolderCreated() throws Exception {
+    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
+    var displayName = "foo";
+    var description = String.format("This is folder %s", displayName);
+
+    ApiFolder folder = createFolder(workspaceId, displayName, description, null);
+
+    assertEquals(displayName, folder.getDisplayName());
+    assertEquals(description, folder.getDescription());
+    assertNotNull(folder.getId());
+    assertNull(folder.getParentFolderId());
+  }
+
+  @Test
+  public void createFolder_doesNotHaveWriteAccess_throws403() throws Exception {
+    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
+    doThrow(new ForbiddenException("User has no write access"))
+        .when(mockSamService)
+        .checkAuthz(
+            any(AuthenticatedUserRequest.class),
+            eq(SamConstants.SamResource.WORKSPACE),
+            anyString(),
+            eq(SamConstants.SamWorkspaceAction.WRITE));
+
+    createFolderExpectCode(
+        workspaceId,
+        "foo",
+        /*description=*/ null,
+        /*parentFolderId=*/ null,
+        HttpStatus.SC_FORBIDDEN);
+  }
+
+  @Test
+  public void createFolder_parentFolderDoesNotExist_throws404() throws Exception {
+    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
+    var displayName = "foo";
+
+    createFolderExpectCode(
+        workspaceId, "foo", /*description=*/ null, UUID.randomUUID(), HttpStatus.SC_NOT_FOUND);
+  }
+
+  @Test
+  public void createFolder_duplicateDisplayName_throws400() throws Exception {
+    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
+    var displayName = "foo";
+
+    // create a top-level folder foo.
+    ApiFolder firstFolder =
+        createFolder(workspaceId, displayName, /*description=*/ null, /*parentFolderId=*/ null);
+    // create another top-level folder foo is not allowed
+    createFolderExpectCode(
+        workspaceId,
+        displayName,
+        /*description=*/ null,
+        /*parentFolderId=*/ null,
+        HttpStatus.SC_BAD_REQUEST);
+
+    // create a second level folder under foo.
+    createFolderExpectCode(
+        workspaceId, displayName, /*description=*/ null, firstFolder.getId(), HttpStatus.SC_OK);
+    createFolderExpectCode(
+        workspaceId,
+        displayName,
+        /*description=*/ null,
+        firstFolder.getId(),
+        HttpStatus.SC_BAD_REQUEST);
+  }
+
+  @Test
+  public void createFolder_duplicateNameAtDifferentFolderLevel_succeeds() throws Exception {
```

Comment:
Delete this test, since it's identical to lines 112-127? (This is where having one test do multiple things makes things easier. Instead of hundreds of lines of tests doing overlapping things, you have one test that's < 50 lines.)

Reply:
this was added by your previous suggestions tho.
I originally just have test for foo/foo. And this one is added for foo/copycat, bar/copycat. 

I changed the test to foo/foo and bar/foo to merge the two tests.